### PR TITLE
Delete object using go client library 

### PR DESCRIPTION
**Pull request by Tulsishah**
Delete Object Using Go Client Library

### CODE REVIEW

File: internal/storage/bucket_handle.go [hunk 1]
```diff
@@ -23,3 +26,25 @@ type bucketHandle struct {
...
+	}
+	// Putting condition that the object's MetaGeneration should match the requested MetaGeneration for deletion to occur.
+	if req.MetaGenerationPrecondition != nil && *req.MetaGenerationPrecondition != 0 {
+		obj = obj.If(storage.Conditions{MetagenerationMatch: *req.MetaGenerationPrecondition})
+	}
+
+	// Deleting object through Go Storage Client.
+	err = obj.Delete(ctx)
+	if err != nil {
+		err = fmt.Errorf("Error in deleting the object through Go storage client: %v", err)
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle.go:45]: when the delete fails what error message do we get. If it says error in deleting the object, then there is no point in adding error message again.
Please check.

Comment (Tulsishah) [reply internal/storage/bucket_handle.go]: resolved

File: internal/storage/bucket_handle_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		Host: "127.0.0.1",
+		Port: 8081,
+	})
+	AssertEq(nil, err)
+}
+func (t *BucketHandleTest) TearDown() {
+	t.fakeStorageServer.Stop()
+}
+func (t *BucketHandleTest) TestDeleteObjectMethodWithValidObject() {
+	fakeClient := t.fakeStorageServer.Client()
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle_test.go:63]: Prince created a seperate file with methods for creating server and object. please refer to those and update the cl.

File: internal/storage/bucket_handle.go [hunk 3]
```diff
@@ -23,3 +25,24 @@ type bucketHandle struct {
...
+		obj = obj.If(storage.Conditions{MetagenerationMatch: *req.MetaGenerationPrecondition})
+	}
+
+	// Deleting object through Go Storage Client.
+	err = obj.Delete(ctx)
+	if err != nil {
+		return
+	}
+
+	return
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle.go:47]: if err is nil or not, you are returning. hence remove line 43-46

**Review by vadlakondaswetha — CHANGES_REQUESTED** [review]

File: internal/storage/bucket_handle.go (hunk 3 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle.go:77]: resolved

File: internal/storage/bucket_handle_test.go [hunk 4]
```diff
@@ -138,3 +139,26 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
 	AssertNe(nil, err)
 	AssertEq(nil, rc)
 }
+
+func (t *BucketHandleTest) TestDeleteObjectMethodWithValidObject() {
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle_test.go:143]: can you testcases which tests the generation and metageneration conditions also.

Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go:143]: Added

File: internal/storage/bucket_handle_test.go [hunk 5]
```diff
@@ -138,3 +139,68 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
...
+		&gcs.DeleteObjectRequest{
+			Name:                       missingObjectName,
+			Generation:                 0,
+			MetaGenerationPrecondition: nil,
+		})
+
+	err_expected := fmt.Errorf("storage: object doesn't exist")
+	AssertEq(err_expected.Error(), error.Error())
+}
+func (t *BucketHandleTest) TestDeleteObjectMethodWithValidGeneration() {
```
Comment (raj-prince) [inline internal/storage/bucket_handle_test.go:165]: Please add a separate line in b/w the methods, this improves the readability.

Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go:166]: fixed

File: internal/storage/bucket_handle_test.go [hunk 6]
```diff
@@ -138,3 +139,68 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
...
+
+func (t *BucketHandleTest) TestDeleteObjectMethodWithMissingObject() {
+	error := t.bucketHandle.DeleteObject(context.Background(),
+		&gcs.DeleteObjectRequest{
+			Name:                       missingObjectName,
+			Generation:                 0,
+			MetaGenerationPrecondition: nil,
+		})
+
+	err_expected := fmt.Errorf("storage: object doesn't exist")
```
Comment (raj-prince) [inline internal/storage/bucket_handle_test.go:162]: Why are we using fmt.Errorf if we are not formatting anything? We can use errors package to create a new error.
ex,
err := errors.New("error string")

File: internal/storage/bucket_handle_test.go [hunk 7]
```diff
@@ -138,3 +139,68 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
...
+		&gcs.DeleteObjectRequest{
+			Name:                       TestObjectName,
+			Generation:                 0,
+			MetaGenerationPrecondition: nil,
+		})
+
+	AssertEq(nil, error)
+}
+func (t *BucketHandleTest) TestDeleteObjectMethodWithInValidMetaGeneration() {
+	var metagenration int64 = 222
```
Comment (raj-prince) [inline internal/storage/bucket_handle_test.go:200]: Please rename metageneration -> metaGeneration

File: internal/storage/bucket_handle_test.go (hunk 6 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go]: fixed

File: internal/storage/bucket_handle_test.go (hunk 7 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go:200]: fixed

**Review by raj-prince — CHANGES_REQUESTED** [review]
Minor comments.

File: internal/storage/bucket_handle.go [hunk 8]
```diff
@@ -60,3 +60,20 @@ func (bh *bucketHandle) NewReader(
 	rc = io.NopCloser(r)
 	return
 }
+func (b *bucketHandle) DeleteObject(ctx context.Context, req *gcs.DeleteObjectRequest) (err error) {
```
[the remark this answers is not in the document]
Comment (Tulsishah) [reply internal/storage/bucket_handle.go:63]: fixed

File: internal/storage/bucket_handle_test.go [hunk 9]
```diff
@@ -138,3 +139,30 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
...
+
+func (t *BucketHandleTest) TestDeleteObjectMethodWithMissingObject() {
+	error := t.bucketHandle.DeleteObject(context.Background(),
+		&gcs.DeleteObjectRequest{
+			Name:                       missingObjectName,
+			Generation:                 0,
+			MetaGenerationPrecondition: nil,
+		})
+
+	err_expected := errors.New("storage: object doesn't exist")
```
Comment (raj-prince) [inline internal/storage/bucket_handle_test.go:162]: We don't need to create the error, if we only comparing the string. Either we can compare directly or can keep an expected_error_message string.

File: internal/storage/bucket_handle_test.go [hunk 10]
```diff
@@ -138,3 +139,30 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
...
+	AssertEq(err_expected.Error(), error.Error())
+}
+
+//Unit tests for Invalid Generation and Metagenaration are not written because
```
Comment (raj-prince) [inline internal/storage/bucket_handle_test.go:166]: >> Space b/w "//" and word.
>> Invalid -> invalid
>> Please use proper punctuation in the comment.
>> Deletemethod -> Delete-method

**Review by raj-prince — APPROVED** [review]
Please incorporate the suggested changes.

File: internal/storage/bucket_handle_test.go (hunk 9 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go]: fixed

File: internal/storage/bucket_handle_test.go (hunk 10 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go]: fixed

File: internal/storage/bucket_handle_test.go [hunk 11]
```diff
@@ -138,3 +138,29 @@ func (t *BucketHandleTest) TestNewReaderMethodWithInvalidGeneration() {
...
+		&gcs.DeleteObjectRequest{
+			Name:                       missingObjectName,
+			Generation:                 0,
+			MetaGenerationPrecondition: nil,
+		})
+
+	AssertEq("storage: object doesn't exist", error.Error())
+}
+
+// Delete-method of the fake server object is not returning an error for invalid Generation and MetaGenerationPrecondition.
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle_test.go:164]: Instead i would suggest rephrasing it to FakeGCSServer is not handling generation and metageneration checks for Delete flow. 
And add this comment at line no: 27.

also link looks weird. Use: https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/vendor/github.com/fsouza/fake-gcs-server/fakestorage/object.go#L515

Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go]: fixed

File: internal/storage/bucket_handle_test.go [hunk 12]
```diff
@@ -25,6 +25,10 @@ import (
 
 const missingObjectName string = "test/foo"
 
+// FakeGCSServer is not handling generation and metageneration checks for Delete flow.
+// Hence, we are not writing the test for this test case.
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle_test.go:29]: we are not writing tests for these flows.

**Review by vadlakondaswetha — APPROVED** [review]

File: internal/storage/bucket_handle_test.go (hunk 12 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go]: fixed

**Pull request merged**